src/libraries/System.ComponentModel.TypeConverter/tests/ConcurrentTypeDescriptorTests.cs

```diff
@@ -0,0 +1,110 @@
+        {
+            var properties = TypeDescriptor.GetProperties(instance);
+            Thread.Sleep(10);
+            if (properties.Count > 0)
+            {
+                Error = true;
+            }
+        }
+
+        [Fact]
+        public void GetProperties_ReturnsExpected()
+        {
+            const int Timeout = 60000;
+            int concurrentCount = Environment.ProcessorCount * 2;
+
+            using var finished = new CountdownEvent(concurrentCount);
+
+            var instances = new TypeWithProperty[concurrentCount];
+            for (int i = 0; i < concurrentCount; i++)
+            {
+                instances[i] = new TypeWithProperty();
+            }
+
+            for (int i = 0; i < concurrentCount; i++)
+            {
+                int i2 = i;
+                new Thread(() =>
+                {
+                    ConcurrentTest(instances[i2]);
+                    finished.Signal();
+                }).Start();
+            }
+
+            finished.Wait(Timeout);
+
+            if (finished.CurrentCount != 0)
+            {
+                Assert.Fail("Timeout. Possible deadlock.");
+            }
+            else
+            {
+                Assert.False(Error, "Fallback type descriptor is used. Possible race condition.");
+            }
+        }
+
+        public sealed class EmptyPropertiesTypeProvider : TypeDescriptionProvider
+        {
+            private sealed class EmptyPropertyListDescriptor : ICustomTypeDescriptor
+            {
+                public AttributeCollection GetAttributes() => AttributeCollection.Empty;
+
+                public string? GetClassName() => null;
+
+                public string? GetComponentName() => null;
+
+                public TypeConverter? GetConverter() => new TypeConverter();
+
+                public EventDescriptor? GetDefaultEvent() => null;
+
+                public PropertyDescriptor? GetDefaultProperty() => null;
+
+                public object? GetEditor(Type editorBaseType) => null;
+
+                public EventDescriptorCollection GetEvents() => EventDescriptorCollection.Empty;
+
+                public EventDescriptorCollection GetEvents(Attribute[]? attributes) => GetEvents();
+
+                public PropertyDescriptorCollection GetProperties() => PropertyDescriptorCollection.Empty;
+
+                public PropertyDescriptorCollection GetProperties(Attribute[]? attributes) => GetProperties();
+
+                public object? GetPropertyOwner(PropertyDescriptor? pd) => null;
+            }
+            public override ICustomTypeDescriptor GetTypeDescriptor(Type objectType, object? instance)
+            {
+                return new EmptyPropertyListDescriptor();
+            }
+        }
+
+        [TypeDescriptionProvider(typeof(EmptyPropertiesTypeProvider))]
+        public sealed class TypeWithProperty
+        {
+            public int OneProperty { get; set; }
+        }
+    }
+}
```

Comment:
Do you think the tests at https://github.com/dotnet/runtime/pull/85156/files#diff-13f56d055c80679284090c59b45d396ab44378385c1f7ad0b597eec15fa5cf0fR813-R909 add any value here?

FWIW locally I added those tests to this file, and ran them in loop for 200 times are no issues were found.

Reply:
The other tests are not fundamentally different - but I add them if they produce errors on your computer.

Both mine and the other tests failed on my PC. Can the issue be platform-specific?

<details>
<summary>test log</summary>


````
✘ System.ComponentModel.Tests.TypeDescriptorTests.GetConverterWithAddProvider_ByMultithread_Success(typeForGetConverter: typeof(System.ComponentModel.Tests.TypeDescriptorTests+MyClass), expectedConverterType: typeof(System.ComponentModel.Tests.TypeDescriptorTests+MyTypeConverter))​5ms
Error:
Assert.All() Failure: 1 out of 200 items in the collection did not pass.
[2]: Item: System.ComponentModel.TypeConverter
     Xunit.Sdk.IsTypeException: Assert.IsType() Failure
     Expected: System.ComponentModel.Tests.TypeDescriptorTests+MyTypeConverter
     Actual:   System.ComponentModel.TypeConverter
        at Xunit.Assert.IsType(Type expectedType, Object object) in /_/src/Microsoft.DotNet.XUnitAssert/src/TypeAsserts.cs:line 140
        at System.ComponentModel.Tests.TypeDescriptorTests.<>c__DisplayClass80_0.<GetConverterWithAddProvider_ByMultithread_Success>b__0(TypeConverter currentConverter) in D:\repos\karakasa\runtime\src\libraries\System.ComponentModel.TypeConverter\tests\TypeDescriptorTests.cs:line 1371
        at Xunit.Assert.<>c__DisplayClass11_0`1.<All>b__0(T item, Int32 index) in /_/src/Microsoft.DotNet.XUnitAssert/src/CollectionAsserts.cs:line 40
        at Xunit.Assert.All[T](IEnumerable`1 collection, Action`2 action) in /_/src/Microsoft.DotNet.XUnitAssert/src/CollectionAsserts.cs:line 69

Stack trace:
   at System.ComponentModel.Tests.TypeDescriptorTests.GetConverterWithAddProvider_ByMultithread_Success(Type typeForGetConverter, Type expectedConverterType) in D:\repos\karakasa\runtime\src\libraries\System.ComponentModel.TypeConverter\tests\TypeDescriptorTests.cs:line 1370
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_0(Object state) in D:\repos\karakasa\runtime\src\libraries\System.Private.CoreLib\src\System\Threading\Tasks\Task.cs:line 1914

✘ System.ComponentModel.Tests.TypeDescriptorTests.GetConverterWithAddProvider_ByMultithread_Success(typeForGetConverter: typeof(System.ComponentModel.Tests.TypeDescriptorTests+MyInheritedClassWithCustomTypeDescriptionProvider), expectedConverterType: typeof(System.ComponentModel.Tests.TypeDescriptorTests+MyInheritedClassWithCustomTypeDescriptionProviderConverter))​2ms
Error:
Assert.All() Failure: 11 out of 200 items in the collection did not pass.
[11]: Item: System.ComponentModel.Tests.TypeDescriptorTests+MyTypeConverter
      Xunit.Sdk.IsTypeException: Assert.IsType() Failure
      Expected: System.ComponentModel.Tests.TypeDescriptorTests+MyInheritedClassWithCustomTypeDescriptionProviderConverter
      Actual:   System.ComponentModel.Tests.TypeDescriptorTests+MyTypeConverter
         at Xunit.Assert.IsType(Type expectedType, Object object) in /_/src/Microsoft.DotNet.XUnitAssert/src/TypeAsserts.cs:line 140
         at System.ComponentModel.Tests.TypeDescriptorTests.<>c__DisplayClass80_0.<GetConverterWithAddProvider_ByMultithread_Success>b__0(TypeConverter currentConverter) in D:\repos\karakasa\runtime\src\libraries\System.ComponentModel.TypeConverter\tests\TypeDescriptorTests.cs:line 1371
         at Xunit.Assert.<>c__DisplayClass11_0`1.<All>b__0(T item, Int32 index) in /_/src/Microsoft.DotNet.XUnitAssert/src/CollectionAsserts.cs:line 40
         at Xunit.Assert.All[T](IEnumerable`1 collection, Action`2 action) in /_/src/Microsoft.DotNet.XUnitAssert/src/CollectionAsserts.cs:line 69
[10]: Item: System.ComponentModel.Tests.TypeDescriptorTests+MyTypeConverter
      Xunit.Sdk.IsTypeException: Assert.IsType() Failure
      Expected: System.ComponentModel.Tests.TypeDescriptorTests+MyInheritedClassWithCustomTypeDescriptionProviderConverter
      Actual:   System.ComponentModel.Tests.TypeDescriptorTests+MyTypeConverter
         at Xunit.Assert.IsType(Type expectedType, Object object) in /_/src/Microsoft.DotNet.XUnitAssert/src/TypeAsserts.cs:line 140
         at System.ComponentModel.Tests.TypeDescriptorTests.<>c__DisplayClass80_0.<GetConverterWithAddProvider_ByMultithread_Success>b__0(TypeConverter currentConverter) in D:\repos\karakasa\runtime\src\libraries\System.ComponentModel.TypeConverter\tests\TypeDescriptorTests.cs:line 1371
         at Xunit.Assert.<>c__DisplayClass11_0`1.<All>b__0(T item, Int32 index) in /_/src/Microsoft.DotNet.XUnitAssert/src/CollectionAsserts.cs:line 40
         at Xunit.Assert.All[T](IEnumerable`1 collection, Action`2 action) in /_/src/Microsoft.DotNet.XUnitAssert/src/CollectionAsserts.cs:line 69
[9]: Item: System.ComponentModel.Tests.TypeDescriptorTests+MyTypeConverter
     Xunit.Sdk.IsTypeException: Assert.IsType() Failure
     Expected: System.ComponentModel.Tests.TypeDescriptorTests+MyInheritedClassWithCustomTypeDescriptionProviderConverter
     Actual:   System.ComponentModel.Tests.TypeDescriptorTests+MyTypeConverter
        at Xunit.Assert.IsType(Type expectedType, Object object) in /_/src/Microsoft.DotNet.XUnitAssert/src/TypeAsserts.cs:line 140
        at System.ComponentModel.Tests.TypeDescriptorTests.<>c__DisplayClass80_0.<GetConverterWithAddProvider_ByMultithread_Success>b__0(TypeConverter currentConverter) in D:\repos\karakasa\runtime\src\libraries\System.ComponentModel.TypeConverter\tests\TypeDescriptorTests.cs:line 1371
        at Xunit.Assert.<>c__DisplayClass11_0`1.<All>b__0(T item, Int32 index) in /_/src/Microsoft.DotNet.XUnitAssert/src/CollectionAsserts.cs:line 40
        at Xunit.Assert.All[T](IEnumerable`1 collection, Action`2 action) in /_/src/Microsoft.DotNet.XUnitAssert/src/CollectionAsserts.cs:line 69
[8]: Item: System.ComponentModel.Tests.TypeDescriptorTests+MyTypeConverter
     Xunit.Sdk.IsTypeException: Assert.IsType() Failure
     Expected: System.ComponentModel.Tests.TypeDescriptorTests+MyInheritedClassWithCustomTypeDescriptionProviderConverter
     Actual:   System.ComponentModel.Tests.TypeDescriptorTests+MyTypeConverter
        at Xunit.Assert.IsType(Type expectedType, Object object) in /_/src/Microsoft.DotNet.XUnitAssert/src/TypeAsserts.cs:line 140
        at System.ComponentModel.Tests.TypeDescriptorTests.<>c__DisplayClass80_0.<GetConverterWithAddProvider_ByMultithread_Success>b__0(TypeConverter currentConverter) in D:\repos\karakasa\runtime\src\libraries\System.ComponentModel.TypeConverter\tests\TypeDescriptorTests.cs:line 1371
        at Xunit.Assert.<>c__DisplayClass11_0`1.<All>b__0(T item, Int32 index) in /_/src/Microsoft.DotNet.XUnitAssert/src/CollectionAsserts.cs:line 40
        at Xunit.Assert.All[T](IEnumerable`1 collection, Action`2 action) in /_/src/Microsoft.DotNet.XUnitAssert/src/CollectionAsserts.cs:line 69
[7]: Item: System.ComponentModel.Tests.TypeDescriptorTests+MyTypeConverter
     Xunit.Sdk.IsTypeException: Assert.IsType() Failure
     Expected: System.ComponentModel.Tests.TypeDescriptorTests+MyInheritedClassWithCustomTypeDescriptionProviderConverter
     Actual:   System.ComponentModel.Tests.TypeDescriptorTests+MyTypeConverter
        at Xunit.Assert.IsType(Type expectedType, Object object) in /_/src/Microsoft.DotNet.XUnitAssert/src/TypeAsserts.cs:line 140
        at System.ComponentModel.Tests.TypeDescriptorTests.<>c__DisplayClass80_0.<GetConverterWithAddProvider_ByMultithread_Success>b__0(TypeConverter currentConverter) in D:\repos\karakasa\runtime\src\libraries\System.ComponentModel.TypeConverter\tests\TypeDescriptorTests.cs:line 1371
        at Xunit.Assert.<>c__DisplayClass11_0`1.<All>b__0(T item, Int32 index) in /_/src/Microsoft.DotNet.XUnitAssert/src/CollectionAsserts.cs:line 40
        at Xunit.Assert.All[T](IEnumerable`1 collection, Action`2 action) in /_/src/Microsoft.DotNet.XUnitAssert/src/CollectionAsserts.cs:line 69
[6]: Item: System.ComponentModel.Tests.TypeDescriptorTests+MyTypeConverter
     Xunit.Sdk.IsTypeException: Assert.IsType() Failure
     Expected: System.ComponentModel.Tests.TypeDescriptorTests+MyInheritedClassWithCustomTypeDescriptionProviderConverter
     Actual:   System.ComponentModel.Tests.TypeDescriptorTests+MyTypeConverter
        at Xunit.Assert.IsType(Type expectedType, Object object) in /_/src/Microsoft.DotNet.XUnitAssert/src/TypeAsserts.cs:line 140
        at System.ComponentModel.Tests.TypeDescriptorTests.<>c__DisplayClass80_0.<GetConverterWithAddProvider_ByMultithread_Success>b__0(TypeConverter currentConverter) in D:\repos\karakasa\runtime\src\libraries\System.ComponentModel.TypeConverter\tests\TypeDescriptorTests.cs:line 1371
        at Xunit.Assert.<>c__DisplayClass11_0`1.<All>b__0(T item, Int32 index) in /_/src/Microsoft.DotNet.XUnitAssert/src/CollectionAsserts.cs:line 40
        at Xunit.Assert.All[T](IEnumerable`1 collection, Action`2 action) in /_/src/Microsoft.DotNet.XUnitAssert/src/CollectionAsserts.cs:line 69
[5]: Item: System.ComponentModel.Tests.TypeDescriptorTests+MyTypeConverter
     Xunit.Sdk.IsTypeException: Assert.IsType() Failure
     Expected: System.ComponentModel.Tests.TypeDescriptorTests+MyInheritedClassWithCustomTypeDescriptionProviderConverter
     Actual:   System.ComponentModel.Tests.TypeDescriptorTests+MyTypeConverter
        at Xunit.Assert.IsType(Type expectedType, Object object) in /_/src/Microsoft.DotNet.XUnitAssert/src/TypeAsserts.cs:line 140
        at System.ComponentModel.Tests.TypeDescriptorTests.<>c__DisplayClass80_0.<GetConverterWithAddProvider_ByMultithread_Success>b__0(TypeConverter currentConverter) in D:\repos\karakasa\runtime\src\libraries\System.ComponentModel.TypeConverter\tests\TypeDescriptorTests.cs:line 1371
        at Xunit.Assert.<>c__DisplayClass11_0`1.<All>b__0(T item, Int32 index) in /_/src/Microsoft.DotNet.XUnitAssert/src/CollectionAsserts.cs:line 40
        at Xunit.Assert.All[T](IEnumerable`1 collection, Action`2 action) in /_/src/Microsoft.DotNet.XUnitAssert/src/CollectionAsserts.cs:line 69
[4]: Item: System.ComponentModel.Tests.TypeDescriptorTests+MyTypeConverter
     Xunit.Sdk.IsTypeException: Assert.IsType() Failure
     Expected: System.ComponentModel.Tests.TypeDescriptorTests+MyInheritedClassWithCustomTypeDescriptionProviderConverter
     Actual:   System.ComponentModel.Tests.TypeDescriptorTests+MyTypeConverter
        at Xunit.Assert.IsType(Type expectedType, Object object) in /_/src/Microsoft.DotNet.XUnitAssert/src/TypeAsserts.cs:line 140
        at System.ComponentModel.Tests.TypeDescriptorTests.<>c__DisplayClass80_0.<GetConverterWithAddProvider_ByMultithread_Success>b__0(TypeConverter currentConverter) in D:\repos\karakasa\runtime\src\libraries\System.ComponentModel.TypeConverter\tests\TypeDescriptorTests.cs:line 1371
        at Xunit.Assert.<>c__DisplayClass11_0`1.<All>b__0(T item, Int32 index) in /_/src/Microsoft.DotNet.XUnitAssert/src/CollectionAsserts.cs:line 40
        at Xunit.Assert.All[T](IEnumerable`1 collection, Action`2 action) in /_/src/Microsoft.DotNet.XUnitAssert/src/CollectionAsserts.cs:line 69
[3]: Item: System.ComponentModel.Tests.TypeDescriptorTests+MyTypeConverter
     Xunit.Sdk.IsTypeException: Assert.IsType() Failure
     Expected: System.ComponentModel.Tests.TypeDescriptorTests+MyInheritedClassWithCustomTypeDescriptionProviderConverter
     Actual:   System.ComponentModel.Tests.TypeDescriptorTests+MyTypeConverter
        at Xunit.Assert.IsType(Type expectedType, Object object) in /_/src/Microsoft.DotNet.XUnitAssert/src/TypeAsserts.cs:line 140
        at System.ComponentModel.Tests.TypeDescriptorTests.<>c__DisplayClass80_0.<GetConverterWithAddProvider_ByMultithread_Success>b__0(TypeConverter currentConverter) in D:\repos\karakasa\runtime\src\libraries\System.ComponentModel.TypeConverter\tests\TypeDescriptorTests.cs:line 1371
        at Xunit.Assert.<>c__DisplayClass11_0`1.<All>b__0(T item, Int32 index) in /_/src/Microsoft.DotNet.XUnitAssert/src/CollectionAsserts.cs:line 40
        at Xunit.Assert.All[T](IEnumerable`1 collection, Action`2 action) in /_/src/Microsoft.DotNet.XUnitAssert/src/CollectionAsserts.cs:line 69
[2]: Item: System.ComponentModel.Tests.TypeDescriptorTests+MyTypeConverter
     Xunit.Sdk.IsTypeException: Assert.IsType() Failure
     Expected: System.ComponentModel.Tests.TypeDescriptorTests+MyInheritedClassWithCustomTypeDescriptionProviderConverter
     Actual:   System.ComponentModel.Tests.TypeDescriptorTests+MyTypeConverter
        at Xunit.Assert.IsType(Type expectedType, Object object) in /_/src/Microsoft.DotNet.XUnitAssert/src/TypeAsserts.cs:line 140
        at System.ComponentModel.Tests.TypeDescriptorTests.<>c__DisplayClass80_0.<GetConverterWithAddProvider_ByMultithread_Success>b__0(TypeConverter currentConverter) in D:\repos\karakasa\runtime\src\libraries\System.ComponentModel.TypeConverter\tests\TypeDescriptorTests.cs:line 1371
        at Xunit.Assert.<>c__DisplayClass11_0`1.<All>b__0(T item, Int32 index) in /_/src/Microsoft.DotNet.XUnitAssert/src/CollectionAsserts.cs:line 40
        at Xunit.Assert.All[T](IEnumerable`1 collection, Action`2 action) in /_/src/Microsoft.DotNet.XUnitAssert/src/CollectionAsserts.cs:line 69
[0]: Item: System.ComponentModel.Tests.TypeDescriptorTests+MyTypeConverter
     Xunit.Sdk.IsTypeException: Assert.IsType() Failure
     Expected: System.ComponentModel.Tests.TypeDescriptorTests+MyInheritedClassWithCustomTypeDescriptionProviderConverter
     Actual:   System.ComponentModel.Tests.TypeDescriptorTests+MyTypeConverter
        at Xunit.Assert.IsType(Type expectedType, Object object) in /_/src/Microsoft.DotNet.XUnitAssert/src/TypeAsserts.cs:line 140
        at System.ComponentModel.Tests.TypeDescriptorTests.<>c__DisplayClass80_0.<GetConverterWithAddProvider_ByMultithread_Success>b__0(TypeConverter currentConverter) in D:\repos\karakasa\runtime\src\libraries\System.ComponentModel.TypeConverter\tests\TypeDescriptorTests.cs:line 1371
        at Xunit.Assert.<>c__DisplayClass11_0`1.<All>b__0(T item, Int32 index) in /_/src/Microsoft.DotNet.XUnitAssert/src/CollectionAsserts.cs:line 40
        at Xunit.Assert.All[T](IEnumerable`1 collection, Action`2 action) in /_/src/Microsoft.DotNet.XUnitAssert/src/CollectionAsserts.cs:line 69

Stack trace:
   at System.ComponentModel.Tests.TypeDescriptorTests.GetConverterWithAddProvider_ByMultithread_Success(Type typeForGetConverter, Type expectedConverterType) in D:\repos\karakasa\runtime\src\libraries\System.ComponentModel.TypeConverter\tests\TypeDescriptorTests.cs:line 1370
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_0(Object state) in D:\repos\karakasa\runtime\src\libraries\System.Private.CoreLib\src\System\Threading\Tasks\Task.cs:line 1914

✘ System.ComponentModel.Tests.TypeDescriptorTests.GetProperties_ReturnsExpected​25ms
Error:
Fallback type descriptor is used. Possible race condition.
Expected: False
Actual:   True

Stack trace:
   at System.ComponentModel.Tests.TypeDescriptorTests.GetProperties_ReturnsExpected() in D:\repos\karakasa\runtime\src\libraries\System.ComponentModel.TypeConverter\tests\TypeDescriptorTests.cs:line 1287
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InterpretedInvoke_Method(Object obj, IntPtr* args) in D:\repos\karakasa\runtime\src\coreclr\System.Private.CoreLib\src\System\Reflection\MethodBaseInvoker.CoreCLR.cs:line 36
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr) in D:\repos\karakasa\runtime\src\libraries\System.Private.CoreLib\src\System\Reflection\MethodBaseInvoker.cs:line 57
````

</details>